// @edoxen/vue@0.1.0/dist/EmptyState.vue_vue_type_style_index_0_lang-BEU3_yM2.js
import { defineComponent as x, computed as h, openBlock as t, createElementBlock as o, toDisplayString as a, createCommentVNode as p, renderSlot as $, Fragment as v, createElementVNode as u, renderList as f, unref as C, createBlock as k, withCtx as w, createTextVNode as L } from "vue";
import { actionTypeLabel as g, compareDecisions as N, countryFlag as T } from "edoxen";
const q = {
  key: 0,
  class: "edoxen-action-badge"
}, A = /* @__PURE__ */ x({
  __name: "ActionTypeBadge",
  props: {
    type: {}
  },
  setup(s) {
    const l = s, e = h(() => g(l.type ?? ""));
    return (_, c) => e.value ? (t(), o("span", q, a(e.value), 1)) : p("", !0);
  }
}), B = { class: "edoxen-prev-next" }, S = { key: 1 }, E = /* @__PURE__ */ x({
  __name: "PrevNextNav",
  props: {
    prev: {},
    next: {}
  },
  emits: ["prev", "next"],
  setup(s, { emit: l }) {
    const e = l;
    return (_, c) => (t(), o("nav", B, [
      $(_.$slots, "empty"),
      s.prev ? (t(), o("button", {
        key: 0,
        type: "button",
        class: "edoxen-prev-next__btn",
        onClick: c[0] || (c[0] = (d) => e("prev"))
      }, " ← " + a(s.prev.title || s.prev.id), 1)) : (t(), o("span", S)),
      s.next ? (t(), o("button", {
        key: 2,
        type: "button",
        class: "edoxen-prev-next__btn",
        onClick: c[1] || (c[1] = (d) => e("next"))
      }, a(s.next.title || s.next.id) + " → ", 1)) : p("", !0)
    ]));
  }
}), F = {
  class: "edoxen-resolution-list",
  role: "list"
}, V = ["onClick"], Y = { class: "edoxen-resolution-list__meta" }, D = { class: "edoxen-resolution-list__year" }, j = { class: "edoxen-resolution-list__id" }, P = { class: "edoxen-resolution-list__title" }, R = {
  key: 0,
  class: "edoxen-resolution-list__snippet"
}, z = {
  key: 1,
  class: "edoxen-resolution-list__actions"
}, K = /* @__PURE__ */ x({
  __name: "ResolutionList",
  props: {
    decisions: {},
    selectedYears: {},
    selectedActionTypes: {},
    query: {},
    sort: {}
  },
  emits: ["select"],
  setup(s, { emit: l }) {
    const e = s, _ = l, c = h(() => {
      var y, n;
      let r = e.decisions;
      if ((y = e.selectedYears) != null && y.length && (r = r.filter((i) => e.selectedYears.includes(i.year))), (n = e.selectedActionTypes) != null && n.length) {
        const i = new Set(e.selectedActionTypes);
        r = r.filter(
          (m) => m.actions.some((b) => i.has(b.type ?? ""))
        );
      }
      if (e.query) {
        const i = e.query.toLowerCase();
        r = r.filter(
          (m) => m.title.toLowerCase().includes(i) || m.subject.toLowerCase().includes(i) || m.identifier.toLowerCase().includes(i)
        );
      }
      return [...r].sort(N(e.sort ?? "newest"));
    });
    function d(r) {
      return r.type ?? "";
    }
    return (r, y) => (t(), o(v, null, [
      u("ul", F, [
        (t(!0), o(v, null, f(c.value, (n) => (t(), o("li", {
          key: n.id,
          class: "edoxen-resolution-list__item",
          onClick: (i) => _("select", n.id)
        }, [
          u("div", Y, [
            u("span", D, a(n.year), 1),
            u("span", j, a(n.identifier), 1)
          ]),
          u("h3", P, a(n.title || n.snippet || C(g)(d(n.actions[0]))), 1),
          n.snippet ? (t(), o("p", R, a(n.snippet), 1)) : p("", !0),
          n.actions.length ? (t(), o("div", z, [
            (t(!0), o(v, null, f(n.actions.slice(0, 3), (i) => (t(), k(A, {
              key: d(i),
              type: d(i)
            }, null, 8, ["type"]))), 128))
          ])) : p("", !0)
        ], 8, V))), 128))
      ]),
      c.value.length === 0 ? (t(), k(E, {
        key: 0,
        prev: null,
        next: null
      }, {
        empty: w(() => [...y[0] || (y[0] = [
          L("No resolutions match the current filters.", -1)
        ])]),
        _: 1
      })) : p("", !0)
    ], 64));
  }
}), G = ["aria-label"], M = /* @__PURE__ */ x({
  __name: "CountryFlag",
  props: {
    country: {}
  },
  setup(s) {
    const l = s, e = h(() => T(l.country ?? null));
    return (_, c) => e.value ? (t(), o("span", {
      key: 0,
      class: "edoxen-flag",
      "aria-label": l.country ?? ""
    }, a(e.value), 9, G)) : p("", !0);
  }
}), H = { class: "edoxen-empty" }, O = /* @__PURE__ */ x({
  __name: "EmptyState",
  props: {
    message: {}
  },
  setup(s) {
    return (l, e) => (t(), o("div", H, [
      u("p", null, a(s.message ?? "Nothing to show."), 1)
    ]));
  }
});
export {
  A as _,
  M as a,
  O as b,
  E as c,
  K as d
};
