// @edoxen/vue@0.1.0/dist/useEdoxenProject-DNZxVs6K.js
import { ref as m, computed as u, watch as v } from "vue";
import { buildProject as w } from "edoxen";
const h = Date.now(), g = /* @__PURE__ */ new Map();
function x(t, e) {
  g.set(t, e);
}
function k() {
  g.clear();
}
function y(t) {
  const { url: e, key: n = e, initial: o, onLoad: a } = t, r = m(o), l = m(!1);
  let s = null;
  return { items: r, isLoaded: l, loadData: async () => {
    if (!l.value) {
      if (s) {
        await s;
        return;
      }
      if (g.has(n)) {
        const i = g.get(n);
        r.value = i, a == null || a(i), l.value = !0;
        return;
      }
      s = (async () => {
        try {
          const f = await (await fetch(`/${e}?t=${h}`)).json();
          r.value = f, a == null || a(f), l.value = !0;
        } catch (i) {
          console.error(`Failed to load ${e}`, i), s = null;
        }
      })(), await s;
    }
  } };
}
function p(t) {
  const { by: e, onLoad: n } = t, o = Array.isArray(e) ? e : [e], a = y({
    url: t.url,
    key: t.key,
    initial: [],
    onLoad: n
  });
  return { ...a, all: () => a.items.value, get: (s) => a.items.value.find(
    (c) => o.some((i) => c[i] === s)
  ) };
}
function b(t = "/data/decisions.json") {
  const e = p({
    url: t,
    by: ["id", "urn", "identifier"]
  });
  return {
    items: u(() => e.items.value),
    isLoaded: u(() => e.isLoaded.value),
    loadData: e.loadData,
    get: e.get
  };
}
function C(t = "/data/meetings.json") {
  const e = p({
    url: t,
    by: ["urn", "slug", "identifier"]
  });
  return {
    items: u(() => e.items.value),
    isLoaded: u(() => e.isLoaded.value),
    loadData: e.loadData,
    get: e.get
  };
}
function L(t = "/data/committee.json") {
  const e = y({
    url: t,
    key: "edoxen:committee",
    initial: null
  });
  return {
    committee: u(() => e.items.value),
    isLoaded: u(() => e.isLoaded.value),
    loadData: e.loadData
  };
}
function U(t = 2e3) {
  const e = m(!1);
  let n = null;
  async function o(a) {
    if (!(!a || typeof navigator > "u" || !navigator.clipboard))
      try {
        await navigator.clipboard.writeText(a), e.value = !0, n && clearTimeout(n), n = setTimeout(() => {
          e.value = !1, n = null;
        }, t);
      } catch {
      }
  }
  return { copied: e, copy: o };
}
function A(t, e, n = 1500) {
  const o = m(0);
  let a = !1;
  return v(
    e,
    (r) => {
      if (!r || a) return;
      const l = t.value;
      if (l == null) return;
      a = !0;
      const s = typeof performance < "u" && typeof performance.now == "function" ? performance.now() : Date.now(), c = (i) => {
        const f = Math.max(0, i - s), d = Math.min(1, f / n), D = 1 - Math.pow(1 - d, 3);
        o.value = Math.round(D * l), d < 1 && requestAnimationFrame(c);
      };
      requestAnimationFrame(c);
    },
    { immediate: !0 }
  ), o;
}
function P(t, e) {
  const n = u(() => {
    const o = e.value;
    return o ? t.value.findIndex((a) => a.id === o) : -1;
  });
  return {
    prev: u(() => n.value > 0 ? t.value[n.value - 1] : null),
    next: u(
      () => n.value >= 0 && n.value < t.value.length - 1 ? t.value[n.value + 1] : null
    )
  };
}
function T(t = {}) {
  const e = b(t.decisionsUrl), n = C(t.meetingsUrl), o = L(t.committeeUrl), a = m(null), r = m(!1);
  let l = [], s = [], c = null;
  function i() {
    a.value = w({
      decisions: l,
      meetings: s,
      committee: c
    });
  }
  return v(
    () => e.items.value,
    (d) => {
      l = d, i();
    },
    { immediate: !0 }
  ), v(
    () => n.items.value,
    (d) => {
      s = d, i();
    },
    { immediate: !0 }
  ), v(
    () => o.committee.value,
    (d) => {
      c = d, i();
    },
    { immediate: !0 }
  ), { project: a, loaded: r, loadData: async () => {
    await Promise.all([
      e.loadData(),
      n.loadData(),
      o.loadData()
    ]), r.value = !0;
  } };
}
export {
  p as a,
  L as b,
  y as c,
  A as d,
  b as e,
  T as f,
  C as g,
  P as h,
  x as i,
  k as r,
  U as u
};
