// @edoxen/edoxen@0.1.0/dist/index.js
import h from "node:fs";
import j from "node:path";
import { load as Q } from "js-yaml";
import * as S from "@asciidoctor/core";
const Z = /\.ya?ml$/i;
function B(e) {
  return h.existsSync(e) ? h.readdirSync(e).filter((t) => Z.test(t)).sort() : [];
}
function X(e) {
  return C(e);
}
function C(e) {
  if (e instanceof Date)
    return e.toISOString().slice(0, 10);
  if (Array.isArray(e))
    return e.map(C);
  if (e && typeof e == "object") {
    const t = {};
    for (const [n, r] of Object.entries(e))
      t[n] = C(r);
    return t;
  }
  return e;
}
function m(e) {
  return typeof e == "object" && e !== null && !Array.isArray(e);
}
function T(e) {
  return m(e) ? {
    metadata: m(e.metadata) ? e.metadata : {},
    decisions: Array.isArray(e.decisions) ? e.decisions : []
  } : { metadata: {}, decisions: [] };
}
function M(e) {
  return m(e) ? Array.isArray(e.meetings) ? {
    metadata: m(e.metadata) ? e.metadata : void 0,
    meetings: e.meetings
  } : "meeting_refs" in e || "identifier" in e && !("type" in e) && !("date_range" in e) ? { series: [e], meetings: [] } : { meetings: [e] } : { meetings: [] };
}
async function Oe(e) {
  if (h.statSync(e).isDirectory()) {
    const n = [];
    let r = {};
    for (const o of B(e)) {
      const i = T(p(j.join(e, o)));
      n.push(...i.decisions), (!r || m(r) && Object.keys(r).length === 0) && (r = i.metadata);
    }
    return { metadata: r, decisions: n };
  }
  return T(p(e));
}
async function Re(e) {
  if (h.statSync(e).isDirectory()) {
    const n = [], r = [];
    for (const o of B(e)) {
      const i = M(p(j.join(e, o)));
      n.push(...i.meetings), i.series && r.push(...i.series);
    }
    return { meetings: n, series: r.length ? r : void 0 };
  }
  return M(p(e));
}
async function $e(e) {
  const t = p(e);
  return m(t) && "meeting_refs" in t ? t : null;
}
async function ze(e) {
  return T(p(e));
}
function p(e) {
  return X(Q(h.readFileSync(e, "utf8")));
}
function ee(e) {
  return Array.isArray(e.dates) && e.dates.length > 0 ? e.dates[0].date || e.dates[0].start || "" : e.date || "";
}
function Fe(e, t, n) {
  const r = e.identifier ?? [], o = $(
    Array.isArray(r) ? r : [r]
  ), i = o.replace(/\//g, "-"), a = R(o), l = ee(n), f = l ? l.slice(0, 4) : "", y = e.title ?? (a && Array.isArray(e.actions) && e.actions.length > 0 ? "Acclamation" : ""), g = e.subject ?? "", k = e.message ?? "", v = O(k || g);
  return {
    id: i,
    identifier: o,
    doi: e.doi ?? "",
    urn: e.urn ?? "",
    year: f,
    venue: n.venue || n.general_area || "",
    city: n.city || "",
    source_file: t,
    meeting_urn: "",
    meeting_date: l,
    is_acclamation: a,
    title: y,
    subject: g,
    snippet: v,
    actions: Array.isArray(e.actions) ? e.actions : [],
    considerations: Array.isArray(e.considerations) ? e.considerations : [],
    approvals: Array.isArray(e.approvals) ? e.approvals : [],
    dates: Array.isArray(e.dates) ? e.dates : [],
    categories: Array.isArray(e.categories) ? e.categories : [],
    urls: Array.isArray(e.urls) ? e.urls : [],
    localizations: Array.isArray(e.localizations) ? e.localizations : void 0,
    raw: e
  };
}
function te(e) {
  const t = e.match(/:meeting:([-\w]+)$/);
  return t ? t[1] : "";
}
function Ie(e, t = {}) {
  var r, o;
  const n = e.urn ?? "";
  return {
    urn: n,
    slug: te(n),
    title: e.title ?? "",
    ordinal: typeof e.ordinal == "number" ? e.ordinal : null,
    type: e.type ?? "",
    status: e.status ?? "",
    bodyType: t.bodyTypeFromCommittee ? t.bodyTypeFromCommittee(e.committee ?? "") : null,
    startDate: ((r = e.date_range) == null ? void 0 : r.start) ?? "",
    endDate: ((o = e.date_range) == null ? void 0 : o.end) ?? "",
    committee: e.committee ?? "",
    venues: Array.isArray(e.venues) ? e.venues : [],
    hosts: Array.isArray(e.hosts) ? e.hosts : [],
    officers: Array.isArray(e.officers) ? e.officers : [],
    components: Array.isArray(e.components) ? e.components : [],
    raw: e
  };
}
const ne = [
  [//g, "•"],
  [//g, "‣"],
  [//g, "▸"],
  [//g, " "]
];
function O(e) {
  if (!e || typeof e != "string") return "";
  let t = e;
  for (const [n, r] of ne)
    t = t.replace(n, r);
  return t = t.replace(/\n+/g, " ").replace(/ {2,}/g, " ").trim(), t.length > 200 && (t = t.slice(0, 197) + "..."), t;
}
function R(e) {
  return typeof e == "string" && e.includes("-acclaim-");
}
function Ne(e, t) {
  return e.title ? e.title : t && Array.isArray(e.actions) && e.actions.length > 0 ? "Acclamation" : "";
}
function Ve(e, t) {
  return e.year !== t.year ? e.year < t.year ? 1 : -1 : e.identifier < t.identifier ? -1 : e.identifier > t.identifier ? 1 : 0;
}
function $(e) {
  return (Array.isArray(e) ? e : [e]).map((n) => {
    const r = n.number || "";
    return n.prefix ? `${n.prefix}/${r}` : r;
  }).join(" / ");
}
function re(e) {
  return e ? e.formatted ? e.formatted : [e.prefix, e.given, e.additional, e.family, e.suffix].filter((t) => t != null && String(t).trim() !== "").join(" ") : "";
}
function D(e, t) {
  if (!Array.isArray(e)) return "";
  const n = e.find((r) => r && r.kind === t && r.primary) || e.find((r) => r && r.kind === t);
  return (n == null ? void 0 : n.value) ?? "";
}
function Pe(e) {
  return e ? {
    name: re(e.name),
    organization: e.affiliation ?? "",
    email: D(e.contact_methods, "email"),
    phone: D(e.contact_methods, "phone")
  } : null;
}
function z(e, t, n) {
  if (!Array.isArray(e)) return null;
  const r = e.find((o) => o && o.profile === t);
  return !r || !Array.isArray(r.attributes) ? null : r.attributes.find((o) => o && o.key === n) ?? null;
}
const F = /* @__PURE__ */ new Map();
function oe(e, t) {
  F.set(e, t);
}
function Ye(e) {
  return F.get(e);
}
function ie(e, t, n) {
  const r = {};
  if (!Array.isArray(e)) return r;
  for (const o of Object.keys(n)) {
    const i = z(e, t, o);
    if (!i) {
      r[o] = null;
      continue;
    }
    r[o] = se(i, n[o]);
  }
  return r;
}
function se(e, t) {
  return t === "integer" && typeof e.intValue == "number" ? e.intValue : t === "float" && typeof e.floatValue == "number" ? e.floatValue : t === "boolean" && typeof e.booleanValue == "boolean" ? e.booleanValue : t === "date" && e.dateValue ? e.dateValue : t === "datetime" && e.dateTimeValue ? e.dateTimeValue : t === "string" && e.value ? e.value : null;
}
async function xe(e, t) {
  if ((t.backend ?? "flexsearch") === "none")
    return new ae(e, t);
  const { createFlexSearchBackend: r } = await import("./flexsearch-CKLS7Zld.js");
  return r(e, t);
}
class ae {
  constructor(t, n) {
    this.docs = t, this.opts = n;
  }
  search(t) {
    if (!t) return this.docs;
    const n = t.toLowerCase();
    return this.docs.filter(
      (r) => this.opts.fields.some((o) => String(r[o] ?? "").toLowerCase().includes(n))
    );
  }
}
const I = /^urn:[a-z0-9][a-z0-9-]{0,30}:[^\s]+$/;
function Ke(e) {
  return typeof e == "string" && I.test(e);
}
function ce(e) {
  if (!I.test(e))
    throw new Error(`Invalid URN: ${e}`);
  return e;
}
function Ge(e) {
  return e.replace(/\//g, "-");
}
function He(e) {
  const t = e.match(/^urn:([a-z0-9][a-z0-9-]{0,30}):(.+)$/);
  return t ? { nid: t[1], nss: t[2] } : null;
}
function ue(e) {
  const t = e.match(/:([-\w]+)$/);
  return t ? t[1] : "";
}
function We(e, t) {
  const n = ue(e);
  return t ? `${t.replace(/\/$/, "")}/${n}` : n;
}
function qe(e, t) {
  return ce(`urn:${e}:meeting:${t}`);
}
const N = /^[a-z]{2,3}$/;
function Je(e) {
  return typeof e == "string" && N.test(e);
}
function Qe(e) {
  if (!N.test(e))
    throw new Error(`Invalid locale: ${e}`);
  return e;
}
function Ze(e, t, n = !0) {
  if (!Array.isArray(e) || e.length === 0) return null;
  const r = e.find((o) => o.language_code === t);
  return r || (n ? e[0] ?? null : null);
}
const le = {
  eng: "en",
  fra: "fr",
  deu: "de",
  spa: "es",
  ita: "it",
  jpn: "ja",
  zho: "zh",
  kor: "ko",
  rus: "ru",
  por: "pt",
  nld: "nl",
  swe: "sv",
  pol: "pl",
  tur: "tr",
  ara: "ar",
  heb: "he",
  hin: "hi"
};
function fe(e) {
  return le[e] ?? e;
}
function de(e, t = "en", n = { year: "numeric", month: "long", day: "numeric" }) {
  if (!e) return "";
  const r = typeof e == "string" ? new Date(e) : e;
  if (isNaN(r.getTime())) return "";
  const o = t.length === 3 ? fe(t) : t;
  return new Intl.DateTimeFormat(o, n).format(r);
}
function Xe(e, t = "en") {
  return de(e, t, { month: "short", day: "numeric" });
}
function et(e) {
  if (!/^[a-z][a-z0-9_-]*$/.test(e))
    throw new Error(`Invalid body code: ${e}`);
  return e;
}
const V = /* @__PURE__ */ new Map();
function ye(e, t) {
  V.set(e, t);
}
function E(e) {
  return V.get(e) ?? [];
}
function me(e, t) {
  var r;
  if (!e) return null;
  const n = e.toLowerCase();
  for (const o of E(t))
    if (n.includes(o.code) || n.includes(o.name.toLowerCase())) return o.code;
  return ((r = E(t)[0]) == null ? void 0 : r.code) ?? null;
}
function pe(e, t) {
  if (!e) return null;
  for (const n of E(t))
    if (e.includes(n.urnPrefix)) return n.code;
  return null;
}
let A = null;
async function ge() {
  if (A) return A;
  const { default: e } = await import("ajv");
  return A = new e({ allErrors: !0, strict: !1 }), A;
}
async function P(e, t) {
  const r = (await ge()).compile(t);
  return r(e) ? { valid: !0, errors: [] } : { valid: !1, errors: (r.errors ?? []).map((l) => ({
    path: String(l.instancePath ?? ""),
    message: String(l.message ?? "validation failed")
  })) };
}
async function tt(e) {
  const t = (await import("./edoxen-Ch_1Fxq0.js")).default;
  return P(e, t);
}
async function nt(e) {
  const t = (await import("./meeting-CZ7-CmHm.js")).default;
  return P(e, t);
}
function rt(e) {
  if (e.bodyTypes && ye(e.scope, e.bodyTypes), e.profiles)
    for (const [t, n] of Object.entries(e.profiles))
      oe(t, n);
  return {
    config: e,
    bodyTypeFromCommittee: (t) => me(t, e.scope),
    bodyTypeFromUrn: (t) => pe(t, e.scope),
    pickExtensionAttribute: (t, n) => z(t, e.scope, n),
    decodeProfile: (t, n, r) => ie(t, n, r)
  };
}
const _ = /* @__PURE__ */ new WeakMap(), U = /* @__PURE__ */ new WeakMap();
function ot(e) {
  const t = _.get(e);
  if (t) return t;
  const n = he(e);
  return _.set(e, n), n;
}
function he(e) {
  const t = $(e.identifier), n = t.replace(/\//g, "-"), r = R(t), o = Array.isArray(e.localizations) ? e.localizations[0] : null, i = (o == null ? void 0 : o.title) ?? (r && e.actions && e.actions.length > 0 ? "Acclamation" : ""), a = (o == null ? void 0 : o.subject) ?? "", l = (o == null ? void 0 : o.message) ?? a;
  return {
    id: n,
    identifier: t,
    urn: e.urn ?? "",
    doi: e.doi ?? "",
    year: Ae(e),
    title: i,
    subject: a,
    snippet: O(l),
    isAcclamation: r,
    actions: e.actions ?? [],
    considerations: e.considerations ?? [],
    approvals: e.approvals ?? []
  };
}
function Ae(e) {
  const t = e.dates ?? [];
  for (const n of t) {
    const r = n.date;
    if (typeof r == "string" && r) return r.slice(0, 4);
  }
  return "";
}
function be(e) {
  const t = /* @__PURE__ */ new Set(), n = [];
  for (const r of e.actions) {
    const o = r.type;
    o && !t.has(o) && (t.add(o), n.push(o));
  }
  return n;
}
function it(e, t = 8) {
  const n = /* @__PURE__ */ new Map();
  for (const i of e)
    for (const a of be(i))
      n.set(a, (n.get(a) ?? 0) + 1);
  const r = Array.from(n.keys()).sort(), o = Array.from(n.entries()).sort((i, a) => a[1] - i[1] || i[0].localeCompare(a[0])).slice(0, t).map(([i]) => i);
  return { all: r, top: o };
}
function st(e) {
  const t = /* @__PURE__ */ new Set();
  for (const n of e) n.year && t.add(n.year);
  return Array.from(t).sort((n, r) => r.localeCompare(n));
}
function at(e) {
  if (!e.length) return { earliest: "", latest: "" };
  const t = [];
  for (const n of e) {
    const r = parseInt(n.year, 10);
    Number.isNaN(r) || t.push(r);
  }
  return t.length ? { earliest: String(Math.min(...t)), latest: String(Math.max(...t)) } : { earliest: "", latest: "" };
}
function ct(e) {
  return e === "oldest" ? (t, n) => t.year.localeCompare(n.year) || t.id.localeCompare(n.id) : e === "most_actions" ? (t, n) => n.actions.length - t.actions.length || n.year.localeCompare(t.year) : (t, n) => n.year.localeCompare(t.year) || n.id.localeCompare(t.id);
}
function ut(e, t = {}) {
  const n = e, r = `${t.title ?? ""}`, o = U.get(n);
  if (o && o.opts === r) return o.vm;
  const i = we(e, t);
  return U.set(n, { vm: i, opts: r }), i;
}
function we(e, t) {
  var f, y;
  const n = e.urn ?? "", r = ((f = n.match(/:meeting:([-\w]+)$/)) == null ? void 0 : f[1]) ?? "", o = e.date_range ?? {}, i = o.start ? parseInt(o.start.slice(0, 4), 10) : null, l = ((y = (e.venues ?? [])[0]) == null ? void 0 : y.name) ?? "";
  return {
    urn: n,
    slug: r,
    title: t.title ?? (i ? `Plenary ${i}` : "Meeting"),
    ordinal: typeof e.ordinal == "number" ? e.ordinal : null,
    type: e.type ?? "",
    status: e.status ?? "",
    year: i,
    start: o.start ?? "",
    end: o.end ?? "",
    venue: l,
    committee: e.committee ?? ""
  };
}
function Y(e) {
  return { ok: !0, value: e };
}
function x(e) {
  return { ok: !1, error: e };
}
function lt(e) {
  try {
    return Y(e());
  } catch (t) {
    return x(t instanceof Error ? t : new Error(String(t)));
  }
}
async function ft(e) {
  try {
    return Y(await e());
  } catch (t) {
    return x(t instanceof Error ? t : new Error(String(t)));
  }
}
function dt(e) {
  return e.ok;
}
function yt(e) {
  return !e.ok;
}
function ke(e) {
  const t = e.meeting;
  return (t == null ? void 0 : t.date) || (t == null ? void 0 : t.venue) || "";
}
function ve(e) {
  const t = e.dates ?? [];
  for (const n of t) {
    const r = n.date;
    if (typeof r == "string" && r) return r.slice(0, 4);
  }
  return "";
}
function Ce(e) {
  return e.kind ?? "";
}
function Te(e) {
  const t = e.date_range, n = t == null ? void 0 : t.start;
  if (typeof n != "string" || !n) return null;
  const r = parseInt(n.slice(0, 4), 10);
  return Number.isNaN(r) ? null : r;
}
function mt(e = {}) {
  const t = e.decisions ?? [], n = e.meetings ?? [], r = e.committee ?? null;
  let o = null, i = null, a = null, l = null, f = null, y = null;
  function g() {
    if (o) return o;
    const s = /* @__PURE__ */ new Map();
    for (const c of t) {
      const u = ke(c);
      if (!u) continue;
      const d = s.get(u);
      d ? d.push(c) : s.set(u, [c]);
    }
    return o = s, s;
  }
  function k() {
    if (i) return i;
    const s = /* @__PURE__ */ new Map();
    for (const c of t) {
      const u = ve(c);
      if (!u) continue;
      const d = s.get(u);
      d ? d.push(c) : s.set(u, [c]);
    }
    return i = s, s;
  }
  function v() {
    if (a) return a;
    const s = /* @__PURE__ */ new Map();
    for (const c of t) {
      const u = Ce(c);
      if (!u) continue;
      const d = s.get(u);
      d ? d.push(c) : s.set(u, [c]);
    }
    return a = s, s;
  }
  function W() {
    if (l) return l;
    const s = /* @__PURE__ */ new Map();
    for (const c of n) {
      const u = c.urn;
      u && s.set(u, c);
    }
    return l = s, s;
  }
  function q() {
    if (f) return f;
    const s = /* @__PURE__ */ new Map();
    for (const c of n) {
      const u = Te(c);
      if (u == null) continue;
      const d = s.get(u);
      d ? d.push(c) : s.set(u, [c]);
    }
    return f = s, s;
  }
  function J() {
    if (y) return y;
    const s = /* @__PURE__ */ new Map();
    for (const c of t) {
      const u = c.urn;
      u && s.set(u, c);
    }
    return y = s, s;
  }
  return {
    committee: r,
    decisions: t,
    meetings: n,
    decisionsByMeeting: () => g(),
    decisionsByYear: () => k(),
    decisionsByKind: () => v(),
    meetingByUrn: (s) => W().get(s),
    meetingByYear: (s) => q().get(s) ?? [],
    decisionByUrn: (s) => J().get(s),
    decisionsForMeeting: (s) => g().get(s) ?? []
  };
}
const b = /^[-•●▪▸▹‣⁃]\s*/, w = /^o\s+/;
function Ee(e) {
  if (!e) return "";
  const t = e.split(`
`), n = [];
  let r = 0;
  for (; r < t.length; ) {
    const o = t[r], i = o.trim();
    if (i === ".") {
      r++;
      continue;
    }
    if (b.test(i)) {
      const a = i.replace(b, "");
      for (n.length > 0 && n[n.length - 1].trim() !== "" && n.push(""), n.push("* " + a), r++; r < t.length && t[r].trim() !== "" && t[r].trim() !== "." && !b.test(t[r].trim()) && !w.test(t[r].trim()) && !w.test(t[r].trim()); ) {
        if (r + 1 < t.length && b.test(t[r + 1].trim())) {
          n.push(t[r]), r++;
          continue;
        }
        break;
      }
      continue;
    }
    if (w.test(i) && i.length < 80) {
      const a = i.replace(w, "");
      n.push("** " + a), r++;
      continue;
    }
    n.push(o), r++;
  }
  return n.join(`
`);
}
function pt(e) {
  if (!e) return "";
  const t = S;
  function n(i, a = 0) {
    if (!(a > 3)) {
      if (typeof i == "function") return i;
      if (i && typeof i == "object") {
        const l = i;
        if (typeof l.convert == "function") return l.convert;
        if ("default" in l) {
          const f = n(l.default, a + 1);
          if (f) return f;
        }
      }
    }
  }
  const r = n(t) ?? (typeof t.convert == "function" ? t.convert : void 0);
  if (!r)
    throw new Error(
      `@asciidoctor/core peer dep missing or incompatible shape. Top-level keys: ${Object.keys(S).slice(0, 10).join(", ")}`
    );
  const o = Ee(e);
  return String(
    r(o, {
      standalone: !1,
      safe: "safe",
      attributes: { showtitle: !1 }
    })
  );
}
let K = (e) => e;
function Se(e) {
  K = e;
}
function gt(e, t) {
  return K(e, t);
}
typeof document < "u" && Se((e, t) => {
  if (!t || !e) return e;
  const n = document.createElement("div");
  n.innerText = e;
  const r = n.innerHTML, o = t.replace(/[.*+?^${}()|[\]\\]/g, "\\$&"), i = new RegExp(`(${o})`, "gi");
  return r.replace(i, '<mark class="search-highlight">$1</mark>');
});
const G = {
  germany: "DE",
  "united states": "US",
  usa: "US",
  "united kingdom": "GB",
  uk: "GB",
  china: "CN",
  france: "FR",
  korea: "KR",
  "south korea": "KR",
  japan: "JP",
  italy: "IT",
  switzerland: "CH",
  australia: "AU",
  sweden: "SE",
  norway: "NO",
  canada: "CA",
  portugal: "PT",
  "south africa": "ZA",
  spain: "ES",
  netherlands: "NL",
  "the netherlands": "NL"
};
function L(e) {
  return e.toUpperCase().split("").map((t) => String.fromCodePoint(127462 + t.charCodeAt(0) - 65)).join("");
}
function Me(e) {
  if (!e) return "";
  const t = e.toLowerCase().trim();
  if (/^[a-z]{2}$/.test(t)) return L(t);
  const n = G[t];
  return n ? L(n) : "";
}
function De(e) {
  if (!e) return "";
  const t = e.split(",");
  return Me(t[t.length - 1]);
}
function _e(e) {
  if (!e) return "";
  const t = e.split(","), n = t[t.length - 1].trim().toLowerCase();
  return /^[a-z]{2}$/.test(n) ? n.toUpperCase() : G[n] ?? "";
}
const ht = De, At = _e, H = {
  adopts: "Adopts",
  thanks: "Thanks",
  approves: "Approves",
  decides: "Decides",
  declares: "Declares",
  asks: "Asks",
  invites: "Invites",
  resolves: "Resolves",
  confirms: "Confirms",
  welcomes: "Welcomes",
  recommends: "Recommends",
  requests: "Requests",
  congratulates: "Congratulates",
  instructs: "Instructs",
  urges: "Urges",
  appoints: "Appoints",
  "calls-upon": "Calls Upon",
  encourages: "Encourages",
  affirms: "Affirms",
  elects: "Elects",
  authorizes: "Authorizes",
  charges: "Charges",
  states: "States",
  remarks: "Remarks",
  judges: "Judges",
  sanctions: "Sanctions",
  abrogates: "Abrogates",
  empowers: "Empowers"
};
function bt(e) {
  return e ? H[e] ?? e.charAt(0).toUpperCase() + e.slice(1) : "";
}
const Ue = {
  adopts: "green",
  approves: "green",
  confirms: "green",
  endorses: "green",
  reaffirms: "green",
  ratifies: "green",
  decides: "blue",
  determines: "blue",
  establishes: "blue",
  instructs: "blue",
  directs: "blue",
  mandates: "blue",
  authorizes: "blue",
  appoints: "blue",
  elects: "blue",
  assigns: "blue",
  nominates: "blue",
  empowers: "blue",
  requests: "amber",
  asks: "amber",
  urges: "amber",
  recommends: "amber",
  suggests: "amber",
  proposes: "amber",
  invites: "amber",
  encourages: "amber",
  resolves: "purple",
  declares: "purple",
  states: "purple",
  announces: "purple",
  proclaims: "purple",
  acknowledges: "slate",
  notes: "slate",
  recognises: "slate",
  recognizes: "slate",
  thanks: "slate",
  appreciates: "slate",
  congratulates: "slate",
  welcomes: "slate",
  abrogates: "red",
  repeals: "red",
  revokes: "red",
  rescinds: "red",
  supersedes: "red",
  replaces: "red",
  withdraws: "red",
  rejects: "red",
  denies: "red",
  refuses: "red",
  opposes: "red"
};
function wt(e) {
  return e ? Ue[e] ?? "slate" : "slate";
}
const kt = Object.freeze(H);
export {
  kt as ACTION_TYPE_LABELS,
  wt as actionTypeColor,
  it as actionTypeFacets,
  bt as actionTypeLabel,
  pt as asciidocify,
  me as bodyTypeFromCommittee,
  pe as bodyTypeFromUrn,
  et as buildBodyCode,
  Ge as buildDecisionId,
  Fe as buildDecisionRecord,
  Qe as buildLocale,
  Ie as buildMeetingRecord,
  qe as buildMeetingUrn,
  mt as buildProject,
  xe as buildSearchIndex,
  ce as buildUrn,
  ct as compareDecisions,
  rt as configureEdoxen,
  _e as countryCodeFromVenue,
  Me as countryFlag,
  ie as decodeProfile,
  Ne as deriveDisplayTitle,
  re as displayName,
  x as err,
  De as flagFromVenue,
  de as formatDate,
  Xe as formatDateShort,
  E as getBodyTypes,
  Ye as getProfile,
  gt as highlightText,
  $ as identifierToDisplay,
  R as isAcclamation,
  yt as isErr,
  Je as isLocale,
  dt as isOk,
  Ke as isUrn,
  fe as iso6393To6391,
  ze as loadCollection,
  Oe as loadDecisions,
  $e as loadMeetingSeries,
  Re as loadMeetings,
  O as normalizeSnippet,
  Y as ok,
  He as parseUrn,
  z as pickExtensionAttribute,
  Ze as pickLocalization,
  Pe as presentContact,
  ot as presentDecision,
  ut as presentMeeting,
  D as primaryContactMethod,
  ye as registerBodyTypes,
  Se as registerHighlightBackend,
  oe as registerProfile,
  ue as slugFromUrn,
  Ve as sortDecisions,
  lt as tryOk,
  ft as tryOkAsync,
  be as uniqueActionTypes,
  We as urnPath,
  P as validateAgainstSchema,
  tt as validateDecisions,
  nt as validateMeetings,
  At as venueToCountryCode,
  ht as venueToFlag,
  B as yamlDir,
  st as yearFacets,
  at as yearRange
};
